### ISSUE TITLE
Document glyphs overlap flags and stroking.

### ISSUE THREAD

**Issue by bungeman** [MicrosoftDocs/typography-issues#922]
It would be nice for the overlap flag bits to be documented in a way so that if a glyph sets an overlap flag it is requesting that its shape be treated as if the areas described by its closed contours were unioned. As already noted in the spec, for filling this is equivalent to using non-zero winding instead of even-odd, and being careful with any overlap in coverage-counting implementations. This appears to have been the original intent of the bit. However, with a more general definition, this can also affect stroking.

When any font with overlapping contours (often a variable font) is stroked naively by stroking the glyph contours, the outlines can be quite messy with unintended crossings and internal stroking. It is possible to remove the overlaps automatically and produce a non-overlapping path, but this can be quite expensive and best avoided for any non-overlapping glyphs. These overlap bits seem well suited to being interpreted as a request to automatically remove any overlap from the path if at all possible, especially when attempting to stroke or otherwise manipulate the outline when drawing the glyph.

I believe this would mean removing the sentence

> Implementations that always use a non-zero-fill algorithm will ignore this flag.

(or modifying it to note that implementations may ignore the flag for rendering purposes) and adding something along the lines of

> Implementations may interpret this flag as a request to remove overlap and simplify the outline before manipulation of the outline, such as for stroking.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b2a9a6a1-a9b9-0864-8982-322da821389e
* Version Independent ID: a42612e3-dc1d-f980-fe4a-4880d1fd0086
* Content: [glyf - Glyf data table (OpenType 1.9) - Typography](https://docs.microsoft.com/en-us/typography/opentype/spec/glyf)
* Content Source: [typographydocs/opentype/spec/glyf.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/opentype/spec/glyf.md)
* Product: **typography**
* GitHub Login: @PeterCon
* Microsoft Alias: **alib**

**Comment by Lorp** [comment]
Does a font maker ever want overlapping edges preserved when stroking?

**Comment by bungeman** [comment]
Far be it from me to speculate on what any particular artist may or may not want. For instance a font designer may want to mark their font as not being allowed to be stroked, fake bolded, fake obliqued, or any of the other horrors often visited upon their designs, including automatic removal of overlaps. While I suspect that it almost always makes sense to remove any overlap when stroking or doing other path manipulation, it is also a lot of work to do so. Interpreting a may-overlap bit set to 1 as a request to do this extra work means that the many existing glyph outlines which don't overlap won't need this extra expensive processing.

Note that one can interpret this proposal in the opposite way: if a glyph doesn't set the may-overlap bit then implementations probably won't attempt to simplify its outline before stroking, so the glyph will probably be processed faster than if simplification were performed. So it's also possible to think of this as stating that outline simplification is part of the "normal" outline processing, but setting the may-overlap bit to 0 opts-out of it.

**Comment by Lorp** [comment]
I believe outline removal already occurs in Microsoft systems when sending variable font instantiations to printers, in order to avoid failing in even-odd winding rasterizers. Though we can presumably treat this as orthogonal to the stroking issue, I would be interested to get MS’s views on the performance hit of overlap removal, to decide whether it is worth setting overlap bits only when necessary. In some instantiation systems I’ve built, I set bit 6 on all simple glyphs (it doesn’t fix stroking, but it gets around the Apple bug).

Remember also that the WOFF2 format [ignores overlap flags](/google/woff2/issues/117).

**Comment by vlevantovsky** [comment]
> Remember also that the WOFF2 format [ignores overlap flags](https://github.com/google/woff2/issues/117).

The last comment on that issue points to proposed corrections that have already been adopted by the W3C. As of March 10, 2022, [the updated WOFF2](https://www.w3.org/TR/WOFF2/) does _not_ ignore the overlap flags :)

**Comment by Lorp** [comment]
> As of March 10, 2022, [the updated WOFF2](https://www.w3.org/TR/WOFF2/) does _not_ ignore the overlap flags :)

That’s great to hear, thanks for taking care of this.

**Comment by bungeman** [comment]
WOFF2 challenge accepted:
https://gitlab.freedesktop.org/freetype/freetype/-/commit/a960be9e5e9846e1e4e8d2254e0b405fbf84346b
https://github.com/google/woff2/commit/4721483ad780ee2b63cb787bfee4aa64b61a0446
https://chromium-review.googlesource.com/c/chromium/src/+/3561539

Hopefully other implementations are or will be updated. There has been some interest from fontmake to set this bit automatically based on the individual instances fed to it and allowing the user to set it directly. I would also be interested in knowing if other implementations are concerned about the cost of always simplifying (or at least detecting) overlapping.

**Comment by khaledhosny** [comment]
Apart from internal glyph overlap, most stroking implementations fail for cursive fonts (e.g. Arabic or Devanagari or even cursive Latin), so there probably should be a general way to handle this that does not benefit only non-cursive scripts.

**Comment by bungeman** [comment]
The cursive (or just generally connecting fonts) issue for stroking is somewhat different, in that it must be done to the paths on a per-draw basis and cannot be statically determined since it involves more than one glyph and they may be positioned in arbitrary ways. An individual glyph isn't really divisible so it can have only one position or transform applied to it, so the glyph outline can be simplified and cached. For connecting fonts an entire run of glyphs must be drawn or simplified as a whole. Without the context of the other glyphs it isn't possible to know which parts of each outline to "leave out". As a result, I think any solution to the overlapping fonts with alpha will also be a good signal for simplifications of entire glyph runs when looking to outline. But that needs to be signaled at a higher level than individual glyphs. While these are related ideas, this issue is about a signal for internal glyph overlap and the connecting fonts issue is about a signal for external overlap.

**Comment by PeterCon** [comment]
@bungeman As you note, variable fonts often have overlaps, but AFAIK there isn't any rasterizer that requires the bit to be set in variable fonts. Hence, we should assume that there are variable fonts with overlaps but without this bit set. Even if we were to recommend that it be used for the benefit of other processes, like stroking, there'd still be many fonts in the wild that haven't set it but that have overlaps. So, it's not clear to me that it would really be beneficial.

**Comment by apodtele** [comment]
There are two kinds of rasterizers:
- Microsoft, for example, uses 4x4 sampling inside each pixels and counts how many sixteenths are inside of a shape. Overlaps do not mater with this method and naturally there is little incentive to take the OVERLAP flags seriously.
- Apple and FreeType use trapezoid integration inside each pixel walking along the contour. In this method, overlaps produce incorrect coverage integral and they can benefit from the explicit OVERLAP flags to mitigate the issue. 

This is _not_ only about even-odd and non-zero fill rule, but also about smudging at the overlapping edges. See the macOS issue microsoft/cascadia-code#362. FreeType now actively uses the OVERLAP flags to mitigate the smudging issue.

The issue is especially nasty when contour edges overlap and produce consecutive runs of incorrect pixels. Overlaps with simple contour crossings are not that bad: a single incorrect pixel is hardly noticeable at the corner. Even though variable fonts frequently use overlaps, the designers should be educated about the nasty edge smudging. Then they could decide if the OVERLAP flag is appropriate or not. Assuming overlaps in all variable glyphs is a bad idea because mitigation is computationally costly.

In summary, _optional_ OVERLAP flags at the discretion of a font designer is a good thing in both variable and static fonts. They help the trapezoid rasterizers deal with smudged edges.

I want to hope that the conflict of interest between different rasterizers can be resolved in the specifications. See #626 for a proposed language.

Cc: @lemzwerg
